### PR TITLE
Adds free surface gradient forcing for idealized 1D turbulence

### PR DESCRIPTION
**Pull request by pwolfram**
Adds free surface gradient forcing for
 * config_zonal_ssh_grad (x-component for planar)
 * config_meridional_ssh_grad (y-component for planar)

Case is applied in periodic flow in
`ocean/periodic_planar/2.5km/default_gotm` case, which can be
used to validate vertical momentum balances with constant forcing
due to double periodicity, which is useful to indicate potential
instabilities with constant forcing due to unstructured mesh.

Note, for full case comparison against literature (e.g., section 5.1 of https://www.sciencedirect.com/science/article/pii/S1463500319302082), needs to have Generaliized
Ocean Turbulence Model to be incorporated.

### CODE REVIEW

**Comment by pwolfram** [comment]
Code generates the following profile (which is non-physical without an appropriate vertical turbulence model to produce the log-law in the vertical):

velocity_profile

Case follows section 5.1 of https://www.sciencedirect.com/science/article/pii/S1463500319302082

image

**Comment by pwolfram** [comment]
@nairita87, this is a good case that can be used to test your implmentation of z_0 at https://github.com/MPAS-Dev/MPAS-Model/pull/606

**Comment by pwolfram** [comment]
@qingli411, can you please try out to see if this provides a log-law in the vertical and parabolic eddy viscosity using GOTM?  This should work with just using a standard C_d=1e-3, but to match the paper @nairita87 's z_0 bottom drag implementation will be needed.

File: testing_and_setup/compass/ocean/periodic_planar/2.5km/default_gotm/config_init.xml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0"?>
+<config case="init">
+
+	<add_executable source="model" dest="ocean_model"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<option name="config_init_configuration">'periodic_planar'</option>
+		<option name="config_vert_levels">-1</option>
+		<option name="config_periodic_planar_vert_levels">25</option>
```
Comment (qingli411) [inline testing_and_setup/compass/ocean/periodic_planar/2.5km/default_gotm/config_init.xml:9]: Should be `250` to match the test in section 5.1 of [Kärnä, 2020](https://doi.org/10.1016/j.ocemod.2020.101619)

**Comment by nairita87** [comment]
> @pwolfram With the pressure gradient term fixed I was able to run GOTM in this test. I computed the bottom friction velocity using a constant drag coefficient `c_d=1e-3` and assuming a constant bottom roughness length `z_0=1.5e-3` m, both needed by GOTM. However, it stopped after running for about 9 hours. The run using CVMix with a constant viscosity also stopped after 8 hours and the run with both CVMix and GOTM off stopped after 7 hours. Below is the velocity profiles in the three cases at hour 7. Note that for the 'none' case it shows the velocity profile right before it breaks. I'm not sure what is happening at the bottom... Also in all cases the bottom shear is very big -- I think perhaps we need the bottom drag @nairita87 is working on to fix this?
> 

Hi @qingli411 I am looking through the changes now. Will update soon. Can you tell me what is the 'none' case? CVMix and GOTM are switched off? So bottom friction is off?

**Comment by qingli411** [comment]
@nairita87 Yes, both CVMix and GOTM are switched off in the 'none' case (I should have used a better name...). It is basically zero viscosity. In all three cases I didn't change the configuration of the bottom drag. It should all be using `config_use_implicit_bottom_drag = .true.` and `config_implicit_bottom_drag_coeff = 1.0e-3`. The bottom friction velocity I computed is separate and is only used in GOTM to compute the viscosity.

**Comment by nairita87** [comment]
Okay @qingli411 . I have added a bottom friction case with a variable config_use_implicit_bottom_roughness = .true. Do you have access to that? 
I will get back with some updates in a bit

**Comment by qingli411** [comment]
@nairita87 Is that in your PR #606? I think I have the access but haven't looked at the code... I will need to merge your changes into my branch.

**Comment by nairita87** [comment]
@qingli411 yes. Okay let me have a look at this

**Comment by pwolfram** [comment]
@qingli411 and @nairita87, please feel free to take the PR from here and make any edits you'd like.  This case really needs combined with GOTM and z_0 in order to do the full validation and thank you both for digging in. 

I'll leave it to you both as this is my last full week working on this work.  Will be around to help with high-level questions if needed, however.

**Comment by qingli411** [comment]
@pwolfram Here are some updates. Please correct me if I'm wrong @nairita87 . 

- @nairita87 sets up a run with the new bottom drag in https://github.com/MPAS-Dev/MPAS-Model/pull/606 with GOTM. The results look much more reasonable than before. See the figure below. Note that this figure plots the velocity profile at one of the center cells (realized that this test case is periodic in x but not in y) whereas the figure I posted earlier was the velocity profile at one of the boundary cells.
- With the new bottom drag we can stably run the test case for 12 hours. We will try a longer run to see how fast it can approach equilibrium.

image

A few things to do next:

- I don't think GOTM knows the updated bottom drag coefficient and roughness length in the present setup. I will work with @nairita87 to make these changes to the GOTM module and have everything consistent.
- Then we can do a more careful comparison with the case in [Kärnä, 2020](https://doi.org/10.1016/j.ocemod.2020.101619)

**Comment by sbrus89** [comment]
@qingli411, I cherry-picked `853756ef7` onto the current `ocean/coastal` here: https://github.com/sbrus89/MPAS-Model/tree/coastal/const_pressure_turb_channel. I can push it @pwolfram's fork to clean up the conflicts. You can see the differences here: https://github.com/sbrus89/MPAS-Model/pull/4

**Comment by qingli411** [comment]
Thanks, @sbrus89! I will rebase my changes in my GOTM branch to the current ocean/coastal branch and create a PR from there

**Comment by sbrus89** [comment]
@qingli411, that sounds great. It may be easier to cherry-pick the specific commits that add the feature rather than do a full rebase since the current ocean/coastal has diverged so much from where you likely started. But either way works.

**Comment by qingli411** [comment]
Hi @sbrus89, I find the history of the ocean/coastal branch very confusing... I started cherry-picking my changes to include GOTM on top of the ocean/coastal branch this morning. But now the commit history is completely changed... I notice that you did a force-push on ocean/coastal. I wonder what was the change?

**Comment by sbrus89** [comment]
@qingli411, it is very confusing. We are trying to sync up `ocean/coastal` and `ocean/develop` to do the #485. I've had to rebase `ocean/coastal` several times as `ocean/develop` has been updated with performance PRs over the last couple weeks. I rebased it again this morning and should have let you know. Sorry for the confusion!

**Comment by qingli411** [comment]
@sbrus89 OK I see. Thanks! So should I use this version of ocean/coastal or should I wait a little bit?

**Comment by sbrus89** [comment]
@qingli411, Maybe just hold off until #485 gets merged. I think this should be happening within the next day or two. That way we avoid any wasted effort.

**Comment by qingli411** [comment]
@sbrus89 Sounds good!

**Comment by mark-petersen** [comment]
@qingli411, this was originally a PR into `ocean/coastal`, so the history is a bit messed up here. But I think the only relevant commit is https://github.com/MPAS-Dev/MPAS-Model/pull/615/commits/853756ef7ce5fbd09de8d58da78fcc1212293ca7.  Looking at that one, I think it is all already included in #704.  At least, it definitely includes the same `'constant_forced'` version of the pressure gradient. I think the test case additions are in #704 as well, in the directory 
```
testing_and_setup/compass/ocean/periodic_planar/2.5km/default_gotm
```

Please take a look. If you agree, I will close this PR.

**Comment by qingli411** [comment]
@mark-petersen Yes, I think this test case is already included in https://github.com/MPAS-Dev/MPAS-Model/pull/704 and I agree that we can close this PR.

**Pull request closed without merging**